File: ruby/get_recent_prioritization_fees.rb
# frozen_string_literal: true

# ruby get_recent_prioritization_fees.rb

require 'json'
require 'dotenv'

Dotenv.load
rpc_url = ENV['RPC_URL']

# Global Fees
crank_grpf = `curl #{ENV['RPC_URL']} -X POST -H "Content-Type: application/json" -d '{"jsonrpc":"2.0", "id":1, "method":"getRecentPrioritizationFees", "params": [[], {"percentile": 7500}]}'`

crank_slot_fees = {}
all_slot_fees = {}

crank_grpf_parsed = JSON.parse(crank_grpf)
puts crank_grpf_parsed

crank_grpf_parsed['result'].each do |obs|
  # {"prioritization_fee"=>10000, "slot"=>172321287}
  crank_slot_fees[obs['slot']] = obs['prioritizationFee']
end
puts ''

crank_slot_fees.sort.each do |k,v|
  puts "#{k}\t#{v}"
end

# show the maximum value of the prioritization fee
puts "\nMAX: #{crank_slot_fees.values.max}\n\n"

File: js/get_recent_prioritization_fees.mjs
// This script will run getRecentPrioritizationFees for a given account
// and print out the details.
//
// Provide RPC_URL in your .env file to check your connection.
//
// Example Use:
// node get_recent_prioritization_fees.mjs <account-pubkey>

import * as web3 from '@solana/web3.js';
import dotenv from 'dotenv';
dotenv.config();

// Read the signature from the command line
const account_string = process.argv[2];
if(!account_string) {throw new Error('Please provide an account pubkey')};

const account = new web3.PublicKey(account_string);

// Set up web3 client
const connection = new web3.Connection(
  process.env.RPC_URL, 'confirmed'
);

console.log(`${account.toBase58()}`);
// Fetch the transaction
const fees = await connection.getRecentPrioritizationFees({lockedWritableAccounts: [account]});

// Loop through fees and print out the details
for (const fee of fees) {
  console.log(fee);
}

// show highest prioritizationFee in fees
const highestFee = fees.reduce((prev, current) => (prev.prioritizationFee > current.prioritizationFee) ? prev : current);
console.log(`Highest prioritization fee: ${highestFee.prioritizationFee}`);

// show average prioritizationFee in fees
const sum = fees.reduce((prev, current) => prev + current.prioritizationFee, 0);
const avg = sum / fees.length;
console.log(`Average prioritization fee: ${avg}`);

// show median prioritizationFee in fees
const sorted = fees.sort((a, b) => a.prioritizationFee - b.prioritizationFee);
const middle = Math.floor(sorted.length / 2);
const isEven = sorted.length % 2 === 0;
const median = isEven ? (sorted[middle].prioritizationFee + sorted[middle - 1].prioritizationFee) / 2 : sorted[middle].prioritizationFee;
console.log(`Median prioritization fee: ${median}`);

// show 90th percentile prioritizationFee in fees
const percentile = 90;
const percentileIndex = Math.floor(sorted.length * (percentile / 100));
const percentileFee = sorted[percentileIndex].prioritizationFee;
console.log(`${percentile}th percentile prioritization fee: ${percentileFee}`);